core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java

```diff
@@ -423,6 +462,20 @@ default void onSuccess() {
          */
         default void onFailure(final Throwable t) {
         }
+
+    }
+
+    public static class IllegalShardFailureException extends ElasticsearchException {
```

Comment:
The danger of reusing is that if the reused exception is thrown *elsewhere* in the course of action (either today or in the future) of executing the shard failure then upstream callbacks that handle the exception will not necessarily be executing for the right reason.